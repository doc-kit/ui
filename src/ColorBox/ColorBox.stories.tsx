import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { withReadme } from 'storybook-readme';
import ColorBox from './ColorBox';

// tslint:disable-next-line no-var-requires
const readme = require('./README.md');

storiesOf('ColorBox', module)
  .addDecorator(withReadme(readme))
  .add('default', () => (
    <ColorBox name={text('name', 'Red Barn')} hex={text('hex', '#F00000')} />
  ));
