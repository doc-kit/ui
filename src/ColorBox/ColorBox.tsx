/** @jsx jsx */
import { jsx, css } from '@emotion/core';
import { Component } from 'react';
import { FiRefreshCw } from 'react-icons/fi';
import { getRgb, hexToRgb, rgbToHsl } from './utils/color-conversions';
import withStyles from '../toolkit/withStyles';
import defaultStyles from './styles';

interface Props {
  name: string;
  hex: string;
  readonly classes: {
    [propName: string]: object;
  };
  readonly mq: (styles: object) => any;
  readonly styles: object;
  [propName: string]: any;
}

interface State {
  clicks: number;
  value: string;
}

class ColorBox extends Component<Props, State> {
  public static defaultProps = {
    hex: '',
    name: '',
  };

  public static isValidHex = /^#[0-9a-f]{3}(?:[0-9a-f]{3})?$/i;
  public static readonly styles = defaultStyles;

  public readonly state = {
    clicks: 1,
    value: this.props.hex,
  };

  public componentDidUpdate(prevProps: Props) {
    const { hex } = this.props;

    if (hex !== prevProps.hex && ColorBox.isValidHex.test(hex)) {
      this.setState({ value: hex, clicks: 1 });
    }
  }

  public render() {
    const { name, hex, classes, mq, styles, ...restOfProps } = this.props;
    const { value } = this.state;

    const rootClass = css(mq(classes.root));
    const wrapperClass = css(mq(classes.wrapper));
    const triangleClass = css(mq(classes.triangle));
    const textClass = css(mq(classes.text));
    const iconClass = css(mq(classes.icon));
    const colorValueClass = css(mq(classes.colorValue));
    const nameClass = css(mq(classes.name));

    return (
      <div css={rootClass} {...restOfProps}>
        <div css={wrapperClass}>
          <div css={triangleClass} />
          <div css={textClass} onClick={this.cycleValues}>
            <div css={iconClass}>
              <FiRefreshCw />
            </div>
            <div css={colorValueClass}>{value}</div>
            <div css={nameClass}>{name}</div>
          </div>
        </div>
      </div>
    );
  }

  private cycleValues = () => {
    const { hex } = this.props;
    const { clicks } = this.state;
    const clickCycle = clicks % 3;
    const [r, g, b] = getRgb(hex);

    this.setState(
      {
        clicks: clicks + 1,
      },
      () => {
        switch (clickCycle) {
          case 1:
            return this.setState({
              value: hexToRgb(hex),
            });
          case 2:
            return this.setState({
              value: rgbToHsl(r, g, b),
            });
          default:
            return this.setState({
              value: this.props.hex,
            });
        }
      }
    );
  };
}

export default withStyles(defaultStyles)(ColorBox);
