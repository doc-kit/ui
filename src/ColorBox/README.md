# Color Box

Useful to render a gallery of available colors belonging to a palette or scheme.

## Features

- Click / tap on the color value to cycle through other formats of the same value.

## Usage

**Import**:

`import { ColorBox } from '@doc-kits/react';`

**JSX**:

```js
<ColorBox colorName="Red Barn" hexValue="#F00000" />
```

## Props

| Property  | Type   | Required | Default | Description                    |
| --------- | ------ | -------- | ------- | ------------------------------ |
| colorName | string | yes      | null    | A nickname for the color shown |
| hexValue  | string | yes      | null    | A valid hex color value        |

## Styles

The default styles injected into `ColorBox` are shown below.

```js
const styles = {
  wrapper: {
    backgroundColor: 'white',
    boxShadow: '0px 2px 4px 0px rgba(0, 0, 0, 0.25)',
    color: '#4a6271',
    display: 'flex',
    fontSize: '1rem',
    height: ['9.375em', '10.938em', '12.500em'], // 150px, 175px, 200px
    lineHeight: '1.5',
    position: 'relative',
    width: ['9.375em', '10.938em', '12.500em'], // 150px, 175px, 200px
  },
  triangle: {
    borderStyle: 'solid',
    borderWidth: ['0 125px 125px 0', '0 150px 150px 0', '0 170px 170px 0'],
    height: '0',
    position: 'absolute',
    right: '0',
    width: '0',
  },
  text: {
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-end',
    padding: '0.5em',
    zIndex: '5',
  },
  icon: {
    color: '#a8a8a8',
    '& svg': {
      width: ['0.750em', '0.875em'],
      height: ['0.750em', '0.875em'],
    },
  },
  colorValue: {
    color: '#a8a8a8',
    fontSize: ['0.625em', '0.750rem', '0.875rem'],
    marginBottom: '0.857em',
    maxWidth: '11.071em',
    textTransform: 'uppercase',
  },
  name: {
    fontSize: ['0.875rem', '1rem', '1.125rem'],
    fontWeight: '600',
    maxWidth: '9.375em',
  },
};
```

## Notes

- You will get the best visual appearance of the component when the `colorName` is kept to a single line of text at all breakpoints.
