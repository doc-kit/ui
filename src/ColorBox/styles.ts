import { darken } from 'polished';
import { ui, palette } from '../toolkit/theme';

const styles = {
  wrapper: {
    backgroundColor: `${darken(0.0099, 'white')}`,
    boxShadow: '0px 2px 4px 0px rgba(0, 0, 0, 0.25)',
    color: ui.text.color,
    display: 'flex',
    fontSize: '1rem',
    height: ['9.375em', '10.938em', '12.500em'], // 150px, 175px, 200px
    lineHeight: '1.5',
    position: 'relative',
    width: ['9.375em', '10.938em', '12.500em'], // 150px, 175px, 200px
  },
  triangle: {
    borderStyle: 'solid',
    borderWidth: ['0 125px 125px 0', '0 150px 150px 0', '0 170px 170px 0'],
    height: '0',
    position: 'absolute',
    right: '0',
    width: '0',
  },
  text: {
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-end',
    padding: '0.5em',
    zIndex: '5',
  },
  icon: {
    color: palette.grayScale['300'],
    '& svg': {
      width: ['0.750em', '0.875em'],
      height: ['0.750em', '0.875em'],
    },
  },
  colorValue: {
    color: palette.grayScale['300'],
    fontSize: ['0.625em', '0.750rem', '0.875rem'],
    marginBottom: '0.857em',
    maxWidth: '11.071em',
    textTransform: 'uppercase',
  },
  name: {
    fontSize: ['0.875rem', '1rem', '1.125rem'],
    fontWeight: '600',
    maxWidth: '9.375em',
  },
};

export default styles;
