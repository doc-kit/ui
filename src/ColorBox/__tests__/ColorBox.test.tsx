import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ColorBox from '../ColorBox';

jest.mock('../../toolkit/constructTheme', () => {
  return () => ({
    ui: { text: {} },
    mq: () => undefined,
  });
});

const mock = {
  componentDidUpdate: jest.spyOn(ColorBox.prototype, 'componentDidUpdate'),
};

describe('<ColorBox />', () => {
  it('should render with minimum requirements', () => {
    const tree = renderer
      .create(<ColorBox name="Red Barn" hex="#F00000" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call componentDidUpdate if hex prop changed', () => {
    const wrapper = shallow(<ColorBox name="Red Barn" hex="#F00000" />);

    wrapper.setProps({ hex: '#000000' });
    expect(mock.componentDidUpdate).toHaveBeenCalled();
  });

  it('should update the state and display value when cycleValues triggered', () => {
    const wrapper = shallow(<ColorBox name="Red Barn" hex="#F00000" />);

    wrapper
      .find('Styled(div)')
      .at(2)
      .simulate('click');

    expect(
      wrapper
        .find('Styled(div)')
        .at(4)
        .text()
    ).toEqual('rgb(240, 0, 0)');
  });
});
