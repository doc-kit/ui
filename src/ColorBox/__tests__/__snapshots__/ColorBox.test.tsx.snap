// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorBox /> should render with minimum requirements 1`] = `
<div
  className="css-7zhfhb"
>
  <div
    className="css-z3iehq"
    color="#F00000"
  />
  <div
    className="css-7zhfhb"
    onClick={[Function]}
  >
    <div
      className="css-7zhfhb"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="23 4 23 10 17 10"
        />
        <polyline
          points="1 20 1 14 7 14"
        />
        <path
          d="M3.51 9a9 9 0 0 1 14.85-3.36L23 10M1 14l4.64 4.36A9 9 0 0 0 20.49 15"
        />
      </svg>
    </div>
    <div
      className="css-7zhfhb"
    >
      #F00000
    </div>
    <div
      className="css-7zhfhb"
    >
      Red Barn
    </div>
  </div>
</div>
`;
